did-doc-diff/src/index.ts

```diff
@@ -0,0 +1,217 @@
+      "tid": "3j55-hlo-v4g4-22",
+      "diffs": {
+        "3j55-hih-g24c-22": { "a": 1, "b": 2, "c": 3 },
+        "3j55-hiu-hcxs-22": {
+          "prev": "3j55-hih-g24c-22",
+          "patches": [["put", ["d"], 4]],
+          "key": "did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez",
+          "sig": "z58kG77wksPLi8185y6fsJG3NidUjYVBzBKVqSG3T5NQKwrNu9d6aL2UorPRzre4VCvKUASuec53jain2LRaMCtmZ",
+        },
+        "3j55-hk2-pbqk-22": {
+          "prev": "3j55-hiu-hcxs-22",
+          "patches": [["del", ["b"]]],
+          "key": "did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez",
+          "sig": "z45gdspAmr9W1tC2KCBrnugdQSXP7fbTUvxz8yM1QnPPNLTKHs2PWzDYgmJe8RPLhZKrfpVFEsvaCMHkhLkFGfxRo",
+        },
+        "3j55-hki-vs6w-22": {
+          "prev": "3j55-hk2-pbqk-22",
+          "patches": [["put", ["e", "ea"], "each"]],
+          "key": "did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez",
+          "sig": "z66WkmrsuVNiqBT3mhjF8G5MNcgpcKbCvXYinqFL73xYkwbSRvNNE7Mj8ak9gEN5FCkamWgsa82GrG984sJxcJdw8",
+        },
+        "3j55-hl6-3ctt-22": {
+          "prev": "3j55-hk2-pbqk-22",
+          "patches": [["put", ["Don't", "put", "me"], "in"]],
+          "key": "did:key:zDnaeYgdnK7nVH2xNQENrBRJbdCQ1KKjxt5sbykiUGLAh46Ez",
+          "sig": "z5oHy2qCSKR9Z9hXuha6c7qEpUWMzGNeZcj8D5kDAxsWP6P8FN4kmSWHcoLv4q5umiyMWcUk5CePGeh51khRGEgGL",
+        }
+      },
+      "key": "did:key:zDnaesnjCLbxSB2EiJgTZUYCBqHTkuDJDdLbSmaAn2bpqXtuY",
+      "sig": "z4Y3LwMCQcmgFvyGfMmSCa5Qr7vCr7tUUDWnRBQqzZsDpSCKP3ZB1WnQTUVBivruavBsCprhjJ5p4WBR1RTyeYqUd"
+    }
+
+  This makes
+    {
+    id: 'did:aic:zwulnkrxrjkobowd',
+    a: 1,
+    c: 3,
+    d: 4,
+    e: { ea: 'each' },
+    }
+*/
+
+export async function tick_to_diffs(tick: {
+  sig: string
+  key: string
+  [index: string]: any
+}) {
+  if (await validate_sig(tick)) {
+    return tick.diffs
+  } else {
+    return undefined
+  }
+}
+
+export async function tick_from_diffs(
+  diffs: { [index: string]: any },
+  key: crypto.EcdsaKeypair,
+) {
+  // This function dose not validate the diffs remember to do that first
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined // ToDo: handel the case where aic has no diffs
+  }
+
+  const tick = {
+    tid: TID.next().formatted(),
+    diffs: diffs,
+    id:
+      'did:aic:' +
+      (await pid(uint8arrays.fromString(JSON.stringify(diffs[tids[0]])))),
+    key: key.did(),
+    sig: '',
+  }
+  const signed_tick = sign(tick, key)
+  return signed_tick
+}
+
+function authorise_key(doc: { [index: string]: any }, didkey: string) {
+  const valid = Boolean(
+    (doc['adx/account_keys'] && doc['adx/account_keys'].includes(didkey)) ||
+      (doc['adx/recovery_keys'] && doc['adx/recovery_keys'].includes(didkey)),
+  )
+  console.log('authorise_key:', valid)
+  return true
+}
+
+export async function diffs_to_did_doc(diffs: { [index: string]: any }) {
+  if (typeof diffs !== 'object') {
+    return undefined
```

Comment:
`undefined`/`null` are weird in JS/TS. they _kinda_ mean the same thing.

my general rule of thumb is: use `undefined` when something just hasn't been set. use `null` when you're explicitly giving a null result

So I use `undefined` for optional fn parameters or optional values in a record. I use `null` for return values or when I want sometime to _explicitly_ acknowledge the fact that they're not providing a value